R/piping-model.R

```diff
@@ -52,26 +52,22 @@ model.rxModelVars <- model.rxode2
   }
   if (.doAppend) {
     # in pre-pending or appending, lines are only added
-    .lhs <- NULL
-    .rhs <- NULL
+    .lhs <- character()
+    .rhs <- character()
     for (x in modelLines) {
-      .isTilde <- identical(x[[1]], quote(`~`))
-      if (.isTilde ||
-            identical(x[[1]], quote(`=`)) ||
-            identical(x[[1]], quote(`<-`))) {
-        .tmp <- .getVariablesFromExpression(x[[3]], ignorePipe=.isTilde)
-        .tmp <- .tmp[!(.tmp %in% .lhs)]
-        .rhs <- unique(c(.tmp, .rhs))
-        .lhs <- unique(c(.getVariablesFromExpression(x[[2]]),.lhs))
+      .isTilde <- .isEndpoint(x)
+      if (.isTilde || .isAssignment(x)) {
+        .rhs <- unique(c(.getVariablesFromExpression(.getRhs(x), ignorePipe=.isTilde), .rhs))
```

Comment:
`.rhs` should not include expressions in `.lhs`.  This matches `rxode2`'s model variables.  You probably need to add

```r
.rhs <- .rhs[!(.rhs %in .lhs)]
```

Reply:
This only really applies to duplicated variables in the same line, but I would code for it anyway.

Reply:
I have that here so that it would work for most lines.  Does this accomplish what you're thinking of?

https://github.com/nlmixr2/rxode2/blob/5dcc4e0eefb5791318c77168955bd7c09b6f423a/R/piping-model.R#L63

Reply:
I did see that line.  In this case, I believe it doesn't apply to duplicated variables in the line.  Defiantly an edge case.

Reply:
Probably need to think of  a test for this edge case, actually....

Reply:
I think the edge case is when you are:
- Adding a line
- Using a residual component.

Perhaps it is OK to have as is.

Reply:
I will let you decide here.  I guess simply tell me so I am aware (that way I can resolve the conversation)

Reply:
I'm not sure that I understand the edge case here.  I'd prefer to take care of it (and add a test for it, whatever is decided) rather than leave it ambiguous.

As I see those lines, the overall goal is to add new `ini()` parts to the model, if necessary.  And, it does so by:

1. Accumulate all new left-hand-side (LHS) variables and right-hand-side (RHS) variables.
2. Anything on the RHS would be added to the `ini()` if it is not on the LHS.

An example model that should have an issue would be something like the following:

```
uiFun <- function() {
  ini({
    a <- 1
  })
  model({
    b <- a + c
    c <- 2
  })
}
```

The problem is that `c` is not defined when `b` asks for it, but then it is later defined.  I think that models like this should be an error regardless of where they come from (model modification or an original model).  The error would be that in the `b <- a + c` line, `c` must come from the dataset since it is not in `ini()`, and on the `c <- 2` line, information from the dataset would be overwritten.  Is this what you were thinking of?

Reply:
No, the above case should work fine.

I was thinking of the case

```
A <- A+eta
```

This occurs rarely but I think is still supported in nlmixr and rxode2.  There are even tests in rxode2 for this kind of behavior.

However this is still captures by the model variables except when they are on endpoint lines. 

This is where it could break since they arent included in the model variables

Something like

```r
Err ~add(err)
```

Which may still work ok...

Reply:
Besides piping comparible models do not need to be compatible with whatever runs I guess.

For now I will mark it as teaolved because the edge case may not even exist